File: AndroidFirebaseAuthExample/app/src/main/java/com/example/shakil/androidfirebaseauthexample/RegisterActivity.java
package com.example.shakil.androidfirebaseauthexample;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.HashMap;

import dmax.dialog.SpotsDialog;

public class RegisterActivity extends AppCompatActivity {

    EditText edtUserName, edtFullName, edtEmail, edtPassword;
    TextView txt_login;
    Button btnRegister;

    FirebaseAuth auth;
    DatabaseReference reference;

    AlertDialog dialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        edtUserName = findViewById(R.id.edtUsername);
        edtFullName = findViewById(R.id.edtFullname);
        edtEmail = findViewById(R.id.edtEmail);
        edtPassword = findViewById(R.id.edtPassword);
        txt_login = findViewById(R.id.txt_login);
        btnRegister = findViewById(R.id.btnRegister);

        dialog = new SpotsDialog.Builder().setCancelable(false).setContext(this).build();

        auth = FirebaseAuth.getInstance();

        txt_login.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(RegisterActivity.this, LoginActivity.class));
            }
        });
        
        btnRegister.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                dialog.show();

                String strUserName = edtUserName.getText().toString();
                String strFullName = edtFullName.getText().toString();
                String strEmail = edtEmail.getText().toString();
                String strPassword = edtPassword.getText().toString();
                
                if (TextUtils.isEmpty(strUserName) || TextUtils.isEmpty(strFullName) 
                        || TextUtils.isEmpty(strEmail) || TextUtils.isEmpty(strPassword)){
                    Toast.makeText(RegisterActivity.this, "All Fields are required", Toast.LENGTH_SHORT).show();
                }
                else if (strPassword.length() < 6){
                    Toast.makeText(RegisterActivity.this, "Password must be at least 6 characters", Toast.LENGTH_SHORT).show();
                }
                else {
                    register(strUserName, strFullName, strEmail, strPassword);
                }
            }
        });
    }

    private void register(final String strUserName, final String strFullName, String strEmail, String strPassword) {
        auth.createUserWithEmailAndPassword(strEmail, strPassword)
                .addOnCompleteListener(RegisterActivity.this, new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                if (task.isSuccessful()){
                    FirebaseUser firebaseUser = auth.getCurrentUser();
                    String userId = firebaseUser.getUid();

                    reference = FirebaseDatabase.getInstance().getReference().child("Users")
                            .child(userId);

                    HashMap<String, Object> hashMap = new HashMap<>();
                    hashMap.put("id", userId);
                    hashMap.put("userName", strUserName.toLowerCase());
                    hashMap.put("fullName", strFullName);

                    reference.setValue(hashMap).addOnCompleteListener(new OnCompleteListener<Void>() {
                        @Override
                        public void onComplete(@NonNull Task<Void> task) {
                            if (task.isSuccessful()){
                                dialog.dismiss();
                                Intent intent = new Intent(RegisterActivity.this, MainActivity.class);
                                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                                startActivity(intent);
                            }
                        }
                    });
                }
                else {
                    dialog.dismiss();
                    Toast.makeText(RegisterActivity.this, "You can't register with this email or password", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }


}

File: KotlinFirebaseAuthExample/app/src/main/java/com/example/shakil/kotlinfirebaseauthexample/RegisterActivity.kt
package com.example.shakil.kotlinfirebaseauthexample

import android.app.AlertDialog
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.Toast
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import dmax.dialog.SpotsDialog
import kotlinx.android.synthetic.main.activity_login.*
import kotlinx.android.synthetic.main.activity_login.edtEmail
import kotlinx.android.synthetic.main.activity_login.edtPassword
import kotlinx.android.synthetic.main.activity_register.*

class RegisterActivity : AppCompatActivity() {

    lateinit var auth: FirebaseAuth
    lateinit var reference: DatabaseReference
    lateinit var dialog: AlertDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)

        auth = FirebaseAuth.getInstance()

        dialog = SpotsDialog.Builder().setContext(this).setCancelable(false).build()

        txt_login.setOnClickListener(View.OnClickListener {
            startActivity(Intent(this@RegisterActivity, LoginActivity::class.java))
        })

        btnRegister.setOnClickListener{
            dialog.show()

            val str_username = edtUsername.text.toString()
            val str_fullname = edtFullname.text.toString()
            val str_email = edtEmail.text.toString()
            val str_password = edtPassword.text.toString()

            if (TextUtils.isEmpty(str_username) || TextUtils.isEmpty(str_fullname)
                || TextUtils.isEmpty(str_email) || TextUtils.isEmpty(str_password)){
                Toast.makeText(this@RegisterActivity, "All Fields are required", Toast.LENGTH_SHORT).show()
            }
            else if (str_password.length < 6){
                Toast.makeText(this@RegisterActivity, "Password must be at least 6 characters", Toast.LENGTH_SHORT).show()
            }
            else{
                register(str_username, str_fullname, str_email, str_password)
            }
        }
    }

    private fun register(strUsername: String, strFullname: String, strEmail: String, strPassword: String) {
        auth.createUserWithEmailAndPassword(strEmail, strPassword).addOnCompleteListener(this@RegisterActivity){
            task ->
            if (task.isSuccessful){
                val  firebaseUser = auth.currentUser
                val userid = firebaseUser!!.uid

                reference = FirebaseDatabase.getInstance().reference.child("Users").child(userid)

                val hashMap = HashMap<String, Any>()
                hashMap["id"] = userid
                hashMap["username"] = strUsername.toLowerCase()
                hashMap["fullname"] = strFullname

                reference.setValue(hashMap).addOnCompleteListener{task ->
                    if (task.isSuccessful){
                        dialog.dismiss()

                        val intent = Intent(this@RegisterActivity, MainActivity::class.java)
                        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
                        startActivity(intent)
                    }
                }
            }
            else{
                dialog.dismiss()
                Toast.makeText(this@RegisterActivity, "You can't register with this email or password", Toast.LENGTH_SHORT).show()
            }
        }
    }
}
